Allow third party extensions to avoid validating the default JAXB context

### Describe the bug

The change 46f348c85934b914f8d986f4f4535134e4ffba03 by @Sgitario has brought hard times to Quarkus CXF and Camel Quarkus. Basically all our CXF related tests fail, because duplicate JAXB classes seem to a normal situation in CXF. 

Here an example of the stack trace see when running [quarkus-cxf-integration-test-client](https://github.com/quarkiverse/quarkus-cxf/tree/main/integration-tests/ws-security-client): 

```
Caused by: java.lang.IllegalStateException: Failed to configure JAXB context
        at io.quarkus.jaxb.deployment.JaxbProcessor.validateContext(JaxbProcessor.java:395)
        at io.quarkus.jaxb.deployment.JaxbProcessor.setupJaxbContextConfig(JaxbProcessor.java:318)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
        at java.base/java.lang.reflect.Method.invoke(Method.java:578)
        at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
        at io.quarkus.builder.BuildContext.run(BuildContext.java:282)
        at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
        at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
        at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
        at java.base/java.lang.Thread.run(Thread.java:1589)
        at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: org.glassfish.jaxb.runtime.v2.runtime.IllegalAnnotationsException: 1 counts of IllegalAnnotationExceptions
Two classes have the same XML type name "{http://www.w3.org/2005/08/addressing}EndpointReferenceType". Use @XmlType.name and @XmlType.namespace to assign different names to them.
        this problem is related to the following location:
                at org.apache.cxf.ws.addressing.EndpointReferenceType
                at public org.apache.cxf.ws.addressing.EndpointReferenceType org.apache.cxf.ws.addressing.ObjectFactory.createEndpointReferenceType()
                at org.apache.cxf.ws.addressing.ObjectFactory
                at protected java.util.List org.apache.cxf.ws.addressing.ReferenceParametersType.any
                at org.apache.cxf.ws.addressing.ReferenceParametersType
        this problem is related to the following location:
                at jakarta.xml.ws.wsaddressing.W3CEndpointReference

        at org.glassfish.jaxb.runtime.v2.runtime.IllegalAnnotationsException$Builder.check(IllegalAnnotationsException.java:83)
        at org.glassfish.jaxb.runtime.v2.runtime.JAXBContextImpl.getTypeInfoSet(JAXBContextImpl.java:421)
        at org.glassfish.jaxb.runtime.v2.runtime.JAXBContextImpl.<init>(JAXBContextImpl.java:255)
        at org.glassfish.jaxb.runtime.v2.runtime.JAXBContextImpl$JAXBContextBuilder.build(JAXBContextImpl.java:1115)
        at org.glassfish.jaxb.runtime.v2.ContextFactory.createContext(ContextFactory.java:144)
        at org.glassfish.jaxb.runtime.v2.JAXBContextFactory.createContext(JAXBContextFactory.java:44)
        at jakarta.xml.bind.ContextFinder.find(ContextFinder.java:368)
        at jakarta.xml.bind.JAXBContext.newInstance(JAXBContext.java:605)
        at jakarta.xml.bind.JAXBContext.newInstance(JAXBContext.java:546)
        at io.quarkus.jaxb.deployment.JaxbProcessor.validateContext(JaxbProcessor.java:393)
        ... 10 more
```

In this particular case, both conflicting classes come through transitive dependencies of CXF:

* `jakarta.xml.ws.wsaddressing.W3CEndpointReference` comes via

```
io.quarkiverse.cxf:quarkus-cxf-integration-test-client:2.0.0-SNAPSHOT:jar:
        -> io.quarkiverse.cxf:quarkus-cxf:2.0.0-SNAPSHOT:jar:
        -> org.apache.cxf:cxf-rt-frontend-jaxws:4.0.0:jar:
        -> org.apache.cxf:cxf-rt-bindings-soap:4.0.0:jar:
        -> jakarta.jws:jakarta.jws-api:3.0.0:jar
```

* `org.apache.cxf.ws.addressing.EndpointReferenceType` comes via

```
io.quarkiverse.cxf:quarkus-cxf-integration-test-client:2.0.0-SNAPSHOT:jar:
        -> io.quarkiverse.cxf:quarkus-cxf:2.0.0-SNAPSHOT:jar:
        -> org.apache.cxf:cxf-rt-frontend-jaxws:4.0.0:jar:
        -> org.apache.cxf:cxf-core:4.0.0:jar
```

I doubt we can exclude the artifacts and I doubt we can exclude the classes via `quarkus.jaxb.exclude-classes`.

Advising all users of `camel-quarkus-cxf` and `quarkus-cxf` to always have `quarkus.jaxb.validate-jaxb-context=false` in their `application.properties` does not seem to be a good UX.

I think the best way forward would be to find a programmatic way of disabling the JAXB context validation. 

I see three options:

### A. Validate the JAXB context only if some JAXBContext injection point is available in the user application. 

### B. Split quarkus-jaxb into quarkus-jaxb-core and quarkus-jaxb

quarkus-jaxb would contain the config, the properties, the validation and the default context bean

quarkus-jaxb-core would keep all the rest. 

quarkus-cxf would depend only on quarkus-jaxb-core. If quarkus-cxf users need to inject the default JAXB context, they are on their own to add quarkus-jaxb dependency and resolve the conflicts.

### C. Have some sort of `ValidateDefaultJAXBContextBuildItem` that would allow third party extensions to signal to JAXB extension that they do not want the validation.

## Assessment:

I see C as rather problematic. I think it would be hard to find the right semantics (esp. how it would combine with `quarkus.jaxb.validate-jaxb-context=false/true`), keep it transparent enough for end users and even document comprehensibly.

If A is technically possible, we should go with A. What do others think?


This blocks 
* https://github.com/apache/camel-quarkus/issues/4603
* https://github.com/quarkiverse/quarkus-cxf/issues/730

--- a/extensions/jaxb/deployment/pom.xml
+++ b/extensions/jaxb/deployment/pom.xml
@@ -28,7 +28,12 @@
 
         <dependency>
             <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-junit5</artifactId>
+            <artifactId>quarkus-junit5-internal</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.assertj</groupId>
+            <artifactId>assertj-core</artifactId>
             <scope>test</scope>
         </dependency>
     </dependencies>

--- a/extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/FilteredJaxbClassesToBeBoundBuildItem.java
+++ b/extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/FilteredJaxbClassesToBeBoundBuildItem.java
@@ -1,0 +1,70 @@
+package io.quarkus.jaxb.deployment;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import io.quarkus.builder.item.SimpleBuildItem;
+import io.quarkus.jaxb.deployment.utils.JaxbType;
+
+/**
+ * List of classes to be bound in the JAXB context. Aggregates all classes passed via
+ * {@link JaxbClassesToBeBoundBuildItem}. All class names excluded via {@code quarkus.jaxb.exclude-classes} are not
+ * present in this list.
+ */
+public final class FilteredJaxbClassesToBeBoundBuildItem extends SimpleBuildItem {
+
+    private final List<Class<?>> classes;
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    private FilteredJaxbClassesToBeBoundBuildItem(List<Class<?>> classes) {
+        this.classes = classes;
+    }
+
+    public List<Class<?>> getClasses() {
+        return new ArrayList<>(classes);
+    }
+
+    public static class Builder {
+        private final Set<String> classNames = new LinkedHashSet<>();
+        private final Set<String> classNameExcludes = new LinkedHashSet<>();
+
+        public Builder classNameExcludes(Collection<String> classNameExcludes) {
+            for (String className : classNameExcludes) {
+                this.classNameExcludes.add(className);
+            }
+            return this;
+        }
+
+        public Builder classNames(Collection<String> classNames) {
+            for (String className : classNames) {
+                this.classNames.add(className);
+            }
+            return this;
+        }
+
+        public FilteredJaxbClassesToBeBoundBuildItem build() {
+            final List<Class<?>> classes = classNames.stream()
+                    .filter(className -> !this.classNameExcludes.contains(className))
+                    .map(FilteredJaxbClassesToBeBoundBuildItem::getClassByName)
+                    .filter(JaxbType::isValidType)
+                    .collect(Collectors.toList());
+
+            return new FilteredJaxbClassesToBeBoundBuildItem(classes);
+        }
+    }
+
+    private static Class<?> getClassByName(String name) {
+        try {
+            return Class.forName(name, false, Thread.currentThread().getContextClassLoader());
+        } catch (ClassNotFoundException e) {
+            throw new RuntimeException(e);
+        }
+    }
+}

--- a/extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbClassesToBeBoundBuildItem.java
+++ b/extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbClassesToBeBoundBuildItem.java
@@ -1,19 +1,24 @@
 package io.quarkus.jaxb.deployment;
 
+import java.util.ArrayList;
+import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
 
 import io.quarkus.builder.item.MultiBuildItem;
 
 /**
- * List of classes to be bound in the JAXB context.
+ * List of class names to be bound in the JAXB context. Note that some of the class names can be removed via
+ * {@code quarkus.jaxb.exclude-classes}.
+ *
+ * @see FilteredJaxbClassesToBeBoundBuildItem
  */
 public final class JaxbClassesToBeBoundBuildItem extends MultiBuildItem {
 
     private final List<String> classes;
 
     public JaxbClassesToBeBoundBuildItem(List<String> classes) {
-        this.classes = Objects.requireNonNull(classes);
+        this.classes = Objects.requireNonNull(Collections.unmodifiableList(new ArrayList<>(classes)));
     }
 
     public List<String> getClasses() {

--- a/extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java
+++ b/extensions/jaxb/deployment/src/main/java/io/quarkus/jaxb/deployment/JaxbProcessor.java
@@ -1,19 +1,16 @@
 package io.quarkus.jaxb.deployment;
 
-import static io.quarkus.jaxb.deployment.utils.JaxbType.isValidType;
-
 import java.io.IOError;
 import java.io.IOException;
 import java.lang.annotation.Annotation;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
-import java.util.Collection;
-import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
 import java.util.stream.Stream;
 
+import jakarta.enterprise.inject.spi.DeploymentException;
 import jakarta.xml.bind.JAXBContext;
 import jakarta.xml.bind.JAXBException;
 import jakarta.xml.bind.annotation.XmlAccessOrder;
@@ -58,6 +55,9 @@ import org.jboss.jandex.Type;
 import org.jboss.logging.Logger;
 
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
+import io.quarkus.arc.deployment.SynthesisFinishedBuildItem;
+import io.quarkus.arc.processor.BeanInfo;
+import io.quarkus.arc.processor.BeanResolver;
 import io.quarkus.deployment.ApplicationArchive;
 import io.quarkus.deployment.annotations.BuildProducer;
 import io.quarkus.deployment.annotations.BuildStep;
@@ -297,29 +297,62 @@ public class JaxbProcessor {
     }
 
     @BuildStep
-    @Record(ExecutionTime.STATIC_INIT)
-    void setupJaxbContextConfig(JaxbConfig config,
-            List<JaxbClassesToBeBoundBuildItem> classesToBeBoundBuildItems,
-            JaxbContextConfigRecorder jaxbContextConfig) {
-        Set<String> classNamesToBeBound = new HashSet<>();
-        for (JaxbClassesToBeBoundBuildItem classesToBeBoundBuildItem : classesToBeBoundBuildItems) {
-            classNamesToBeBound.addAll(classesToBeBoundBuildItem.getClasses());
-        }
+    FilteredJaxbClassesToBeBoundBuildItem filterBoundClasses(
+            JaxbConfig config,
+            List<JaxbClassesToBeBoundBuildItem> classesToBeBoundBuildItems) {
+
+        FilteredJaxbClassesToBeBoundBuildItem.Builder builder = FilteredJaxbClassesToBeBoundBuildItem.builder();
+        classesToBeBoundBuildItems.stream()
+                .map(JaxbClassesToBeBoundBuildItem::getClasses)
+                .forEach(builder::classNames);
 
         // remove classes that have been excluded by users
         if (config.excludeClasses.isPresent()) {
-            classNamesToBeBound.removeAll(config.excludeClasses.get());
+            builder.classNameExcludes(config.excludeClasses.get());
         }
+        return builder.build();
+    }
 
-        // parse class names to class
-        Set<Class<?>> classes = getAllClassesFromClassNames(classNamesToBeBound);
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
+    void setupJaxbContextConfig(
+            FilteredJaxbClassesToBeBoundBuildItem filteredClassesToBeBound,
+            JaxbContextConfigRecorder jaxbContextConfig) {
+        jaxbContextConfig.addClassesToBeBound(filteredClassesToBeBound.getClasses());
+    }
+
+    @BuildStep
+    @Record(ExecutionTime.STATIC_INIT)
+    void validateDefaultJaxbContext(
+            JaxbConfig config,
+            FilteredJaxbClassesToBeBoundBuildItem filteredClassesToBeBound,
+            SynthesisFinishedBuildItem beanContainerState,
+            JaxbContextConfigRecorder jaxbContextConfig /* Force the build time container to invoke this method */) {
+
         if (config.validateJaxbContext) {
-            // validate the context to fail at build time if it's not valid
-            validateContext(classes);
+            final BeanResolver beanResolver = beanContainerState.getBeanResolver();
+            final Set<BeanInfo> beans = beanResolver
+                    .resolveBeans(Type.create(DotName.createSimple(JAXBContext.class), org.jboss.jandex.Type.Kind.CLASS));
+            if (!beans.isEmpty()) {
+                final BeanInfo bean = beanResolver.resolveAmbiguity(beans);
+                if (bean.isDefaultBean()) {
+                    /*
+                     * Validate the default JAXB context at build time and fail early.
+                     * Do this only if the user application actually requires the default JAXBContext bean
+                     */
+                    try {
+                        JAXBContext.newInstance(filteredClassesToBeBound.getClasses().toArray(new Class[0]));
+                    } catch (JAXBException e) {
+                        /*
+                         * Producing a ValidationErrorBuildItem would perhaps be more natural here,
+                         * but doing so causes a cycle between this and reactive JAXB extension
+                         * Throwing from here works well too
+                         */
+                        throw new DeploymentException("Failed to create or validate the default JAXBContext", e);
+                    }
+                }
+            }
         }
-
-        // register the classes to be used at runtime
-        jaxbContextConfig.addClassesToBeBound(classes);
     }
 
     @BuildStep
@@ -388,31 +421,4 @@ public class JaxbProcessor {
         resourceBundle.produce(new NativeImageResourceBundleBuildItem(bundle));
     }
 
-    private void validateContext(Set<Class<?>> classes) {
-        try {
-            JAXBContext.newInstance(classes.toArray(new Class[0]));
-        } catch (JAXBException e) {
-            throw new IllegalStateException("Failed to configure JAXB context", e);
-        }
-    }
-
-    private Set<Class<?>> getAllClassesFromClassNames(Collection<String> classNames) {
-        Set<Class<?>> classes = new HashSet<>();
-        for (String className : classNames) {
-            Class<?> clazz = getClassByName(className);
-            if (isValidType(clazz)) {
-                classes.add(clazz);
-            }
-        }
-
-        return classes;
-    }
-
-    private Class<?> getClassByName(String name) {
-        try {
-            return Class.forName(name, false, Thread.currentThread().getContextClassLoader());
-        } catch (ClassNotFoundException e) {
-            throw new RuntimeException(e);
-        }
-    }
 }

--- a/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/ConflictingModelClassesMarshalerOnlyTest.java
+++ b/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/ConflictingModelClassesMarshalerOnlyTest.java
@@ -1,0 +1,54 @@
+package io.quarkus.jaxb.deployment;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+
+import jakarta.enterprise.context.control.ActivateRequestContext;
+import jakarta.enterprise.inject.spi.DeploymentException;
+import jakarta.inject.Inject;
+import jakarta.xml.bind.JAXBContext;
+import jakarta.xml.bind.Marshaller;
+
+import org.assertj.core.api.Assertions;
+import org.glassfish.jaxb.core.v2.runtime.IllegalAnnotationException;
+import org.glassfish.jaxb.runtime.v2.runtime.IllegalAnnotationsException;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Make sure that the validation of the default JAXB context fails if there conflicting model classes and there is only
+ * a {@link Marshaller} injection point (which actually requires a {@link JAXBContext} bean to be available too).
+ */
+public class ConflictingModelClassesMarshalerOnlyTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(
+                            io.quarkus.jaxb.deployment.one.Model.class,
+                            io.quarkus.jaxb.deployment.two.Model.class))
+            .assertException(e -> {
+                assertThat(e).isInstanceOf(DeploymentException.class);
+                assertThat(e.getMessage()).isEqualTo("Failed to create or validate the default JAXBContext");
+                Throwable cause = e.getCause();
+                assertThat(cause).isInstanceOf(IllegalAnnotationsException.class);
+                assertThat(cause.getMessage()).isEqualTo("1 counts of IllegalAnnotationExceptions");
+                List<IllegalAnnotationException> errors = ((IllegalAnnotationsException) cause).getErrors();
+                assertThat(errors.size()).isEqualTo(1);
+                assertThat(errors.get(0).getMessage()).contains("Two classes have the same XML type name \"model\"");
+
+            });
+
+    @Inject
+    Marshaller marshaller;
+
+    @Test
+    @ActivateRequestContext
+    public void shouldFail() {
+        Assertions.fail("The application should fail at boot");
+    }
+
+}

--- a/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/ConflictingModelClassesTest.java
+++ b/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/ConflictingModelClassesTest.java
@@ -1,0 +1,53 @@
+package io.quarkus.jaxb.deployment;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+
+import jakarta.enterprise.context.control.ActivateRequestContext;
+import jakarta.enterprise.inject.spi.DeploymentException;
+import jakarta.inject.Inject;
+import jakarta.xml.bind.JAXBContext;
+
+import org.assertj.core.api.Assertions;
+import org.glassfish.jaxb.core.v2.runtime.IllegalAnnotationException;
+import org.glassfish.jaxb.runtime.v2.runtime.IllegalAnnotationsException;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Make sure that the validation of the default JAXB context fails if there conflicting model classes and there actually
+ * is a {@link JAXBContext} injection point.
+ */
+public class ConflictingModelClassesTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(
+                            io.quarkus.jaxb.deployment.one.Model.class,
+                            io.quarkus.jaxb.deployment.two.Model.class))
+            .assertException(e -> {
+                assertThat(e).isInstanceOf(DeploymentException.class);
+                assertThat(e.getMessage()).isEqualTo("Failed to create or validate the default JAXBContext");
+                Throwable cause = e.getCause();
+                assertThat(cause).isInstanceOf(IllegalAnnotationsException.class);
+                assertThat(cause.getMessage()).isEqualTo("1 counts of IllegalAnnotationExceptions");
+                List<IllegalAnnotationException> errors = ((IllegalAnnotationsException) cause).getErrors();
+                assertThat(errors.size()).isEqualTo(1);
+                assertThat(errors.get(0).getMessage()).contains("Two classes have the same XML type name \"model\"");
+
+            });
+
+    @Inject
+    JAXBContext jaxbContext;
+
+    @Test
+    @ActivateRequestContext
+    public void shouldFail() {
+        Assertions.fail("The application should fail at boot");
+    }
+
+}

--- a/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/InjectJaxbContextTest.java
+++ b/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/InjectJaxbContextTest.java
@@ -1,10 +1,10 @@
 package io.quarkus.jaxb.deployment;
 
-import static org.junit.jupiter.api.Assertions.assertEquals;
-import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.assertj.core.api.Assertions.assertThat;
 
 import java.io.StringWriter;
 
+import jakarta.enterprise.context.control.ActivateRequestContext;
 import jakarta.inject.Inject;
 import jakarta.xml.bind.JAXBContext;
 import jakarta.xml.bind.JAXBException;
@@ -12,11 +12,27 @@ import jakarta.xml.bind.Marshaller;
 import jakarta.xml.bind.Unmarshaller;
 
 import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
 
-import io.quarkus.test.junit.QuarkusTest;
+import io.quarkus.test.QuarkusUnitTest;
 
-@QuarkusTest
+/**
+ * Make sure that the default JAXBContext passes the validation thanks to
+ * {@code quarkus.jaxb.exclude-classes=io.quarkus.jaxb.deployment.two.Model} even though there are conflicting classes
+ * in the application.
+ */
 public class InjectJaxbContextTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(
+                            io.quarkus.jaxb.deployment.one.Model.class,
+                            io.quarkus.jaxb.deployment.two.Model.class,
+                            Person.class,
+                            CustomJaxbContextCustomizer.class)
+                    .addPackage("io.quarkus.jaxb.deployment.info"))
+            .overrideConfigKey("quarkus.jaxb.exclude-classes", "io.quarkus.jaxb.deployment.two.Model");
 
     @Inject
     JAXBContext jaxbContext;
@@ -28,13 +44,23 @@ public class InjectJaxbContextTest {
     Unmarshaller unmarshaller;
 
     @Test
+    @ActivateRequestContext
     public void shouldInjectJaxbBeans() {
-        assertNotNull(jaxbContext);
-        assertNotNull(marshaller);
-        assertNotNull(unmarshaller);
+        assertThat(jaxbContext).isNotNull();
+        assertThat(marshaller).isNotNull();
+        assertThat(unmarshaller).isNotNull();
     }
 
     @Test
+    @ActivateRequestContext
+    public void packageInfoLoaded() {
+        /* make sure the package-info.class is present in the test archive */
+        assertThat(io.quarkus.jaxb.deployment.info.Foo.class.getPackage()
+                .getAnnotation(jakarta.xml.bind.annotation.XmlSchema.class)).isNotNull();
+    }
+
+    @Test
+    @ActivateRequestContext
     public void shouldPersonBeInTheJaxbContext() throws JAXBException {
         Person person = new Person();
         person.setFirst("first");
@@ -43,11 +69,11 @@ public class InjectJaxbContextTest {
         StringWriter sw = new StringWriter();
         marshaller.marshal(person, sw);
 
-        assertEquals("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n"
+        assertThat(sw.toString()).isEqualTo("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n"
                 + "<person>\n"
                 + "    <first>first</first>\n"
                 + "    <last>last</last>\n"
-                + "</person>\n", sw.toString());
+                + "</person>\n");
     }
 
 }

--- a/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/NoJaxbContextBeanTest.java
+++ b/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/NoJaxbContextBeanTest.java
@@ -1,0 +1,34 @@
+package io.quarkus.jaxb.deployment;
+
+import jakarta.enterprise.context.control.ActivateRequestContext;
+import jakarta.xml.bind.JAXBContext;
+
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.InstanceHandle;
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * Make sure that the default JAXB context is not validated at build time as long as there is no injection point for it.
+ * Conflicting model classes thus won't make the application fail.
+ */
+public class NoJaxbContextBeanTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(
+                            io.quarkus.jaxb.deployment.one.Model.class,
+                            io.quarkus.jaxb.deployment.two.Model.class));
+
+    @Test
+    @ActivateRequestContext
+    public void noJaxbContext() {
+        InstanceHandle<JAXBContext> contextHandle = Arc.container().instance(JAXBContext.class);
+        Assertions.assertFalse(contextHandle.isAvailable());
+    }
+
+}

--- a/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/UserProvidedJaxbContextTest.java
+++ b/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/UserProvidedJaxbContextTest.java
@@ -1,0 +1,91 @@
+package io.quarkus.jaxb.deployment;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.io.StringWriter;
+
+import jakarta.enterprise.context.control.ActivateRequestContext;
+import jakarta.enterprise.inject.Produces;
+import jakarta.inject.Inject;
+import jakarta.inject.Singleton;
+import jakarta.xml.bind.JAXBContext;
+import jakarta.xml.bind.JAXBException;
+import jakarta.xml.bind.Marshaller;
+import jakarta.xml.bind.Unmarshaller;
+
+import org.assertj.core.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+/**
+ * If the user provides his own {@link JAXBContext} bean then the validation of the default context should not happen
+ * and the presence of conflicting classes, such as {@link io.quarkus.jaxb.deployment.one.Model} and
+ * {@link io.quarkus.jaxb.deployment.two.Model} should not matter.
+ */
+public class UserProvidedJaxbContextTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(
+                            io.quarkus.jaxb.deployment.one.Model.class,
+                            io.quarkus.jaxb.deployment.two.Model.class,
+                            UserProvidedJaxbContextTest.JaxbContextProducer.class));
+
+    @Inject
+    JAXBContext jaxbContext;
+
+    @Inject
+    Marshaller marshaller;
+
+    @Inject
+    Unmarshaller unmarshaller;
+
+    @Test
+    @ActivateRequestContext
+    public void shouldInjectJaxbBeans() {
+        assertThat(jaxbContext).isNotNull();
+        assertThat(marshaller).isNotNull();
+        assertThat(unmarshaller).isNotNull();
+    }
+
+    @Test
+    @ActivateRequestContext
+    public void marshalModelOne() throws JAXBException {
+        io.quarkus.jaxb.deployment.one.Model model = new io.quarkus.jaxb.deployment.one.Model();
+        model.setName1("name1");
+
+        StringWriter sw = new StringWriter();
+        marshaller.marshal(model, sw);
+
+        assertThat(sw.toString()).isEqualTo("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>"
+                + "<model><name1>name1</name1></model>");
+    }
+
+    @Test
+    @ActivateRequestContext
+    public void marshalModelTwo() throws JAXBException {
+        io.quarkus.jaxb.deployment.two.Model model = new io.quarkus.jaxb.deployment.two.Model();
+        model.setName2("name2");
+        Assertions.assertThatExceptionOfType(JAXBException.class)
+                .isThrownBy(() -> marshaller.marshal(model, new StringWriter()))
+                .withMessage("class io.quarkus.jaxb.deployment.two.Model nor any of its super class is known to this context.");
+    }
+
+    public static class JaxbContextProducer {
+
+        @Produces
+        @Singleton
+        JAXBContext produceJaxbContext() {
+            try {
+                return JAXBContext.newInstance(new Class[] { io.quarkus.jaxb.deployment.one.Model.class });
+            } catch (JAXBException e) {
+                throw new RuntimeException("Could not create new JAXBContext", e);
+            }
+        }
+
+    }
+
+}

--- a/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/info/Foo.java
+++ b/extensions/jaxb/deployment/src/test/java/io/quarkus/jaxb/deployment/info/Foo.java
@@ -1,0 +1,4 @@
+package io.quarkus.jaxb.deployment.info;
+
+public class Foo {
+}

--- a/extensions/jaxb/deployment/src/test/resources/application.properties
+++ b/extensions/jaxb/deployment/src/test/resources/application.properties
@@ -1,1 +1,0 @@
-quarkus.jaxb.exclude-classes=io.quarkus.jaxb.deployment.two.Model

